packages/vite/src/node/plugins/define.ts

```diff
@@ -100,9 +99,11 @@ export function definePlugin(config: ResolvedConfig): Plugin {
 
       if (ssr && !isBuild) {
         // ssr + dev, simple replace
-        return code.replace(pattern, (_, match) => {
-          return '' + replacements[match]
-        })
+        return Object.keys(replacements).length
```

Comment:
It looks like the MagicString code below could benefit from this optimization as well, e.g. during SSR builds. Perhaps we can move this check above the if condition, to early return null when possible.

I also think we can optimize the `Object.keys(replacements).length` code, since that creates a new array on every call. Maybe we can return a `null` pattern when there's no replacements.

Reply:
I suggested against the second one with the prev implementation, but a null pattern could work nicely here

Reply:
> It looks like the MagicString code below could benefit from this optimization as well, e.g. during SSR builds. Perhaps we can move this check above the if condition, to early return null when possible.

thanks @bluwy, nice to see your review. I think processNodeEnv is always had value when is not during SSR builds, may not benefit from this optimization. but move this check above the if condition is look good 👍

Reply:
> I also think we can optimize the Object.keys(replacements).length code

@patak-dev suggested

> We don't win much in speed to justify it.

😂 how can I do that?

Reply:
I think patak's comment about mentioned that a null pattern would be fine, so ideally we can check `if (pattern == null) return null`, instead of the `Object.keys` code.

Reply:
oh, it turns out that the pattern value here is a variable 😂. thank again @bluwy